### ISSUE TITLE
[JENKINS-27329] WorkspaceCleanupThread may delete workspaces of running jobs

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-15#9588]
<p>The problem is as described in <a href="https://issues.jenkins.io/browse/JENKINS-4501" title="WorkspaceCleanupThread deletes Matrix project workspaces" class="issue-link" data-issue-key="JENKINS-4501"><del>JENKINS-4501</del></a>. As requested in <a href="https://issues.jenkins.io/browse/JENKINS-4501" title="WorkspaceCleanupThread deletes Matrix project workspaces" class="issue-link" data-issue-key="JENKINS-4501"><del>JENKINS-4501</del></a>, I am creating a new issue as this problem still exists in 1.602.</p>

<p>In short, Jenkins silently and erroneously deletes workspaces on slaves for matrix projects that are not old.</p>

<p>Over the course of the time I've worked with Jenkins this behavior has created literally days of work and waiting on very long running builds that rely on cached workspaces to be manageable. It's cost me more hours again today after restoring jenkins to a new server after a hardware failure. This setting was reset since it exists outside normal recommended backup files and I didn't think to add it when I "fixed" this last time.</p>

<p>Would it not be easier to have hudson.model.WorkspaceCleanupThread.disabled default to true? Having the default behavior be "destroy my data" seems bad, especially with how cheap disk is now. I'm sure when this option was implemented it made a lot of sense, but when I can get a 1TB for $50, it just seems wrong-headed. Let the fallow workspaces lie. I can clean them up if I need to.</p>

<p>If that's not an acceptable solution, could it not be moved to a config location in the Jenkins home? That way we can be relatively sure that the setting will be propagated in backups and not bite someone who thought they solved this problem and had forgotten about it?</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/qhartman.jpg" title="qhartman's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=qhartman">qhartman</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-27329" target="_blank">WorkspaceCleanupThread may delete workspaces of running jobs</a></i></summary>
<i><ul>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>component(s)</b>: core
<li><b>label(s)</b>: 2.121.2-fixed
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2018-06-06T09:43:24+00:00
<li><b>votes</b>: 13
<li><b>watchers</b>: 19
<li><b>imported</b>: 2025-11-16
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>The problem is as described in <a href="https://issues.jenkins.io/browse/JENKINS-4501" title="WorkspaceCleanupThread deletes Matrix project workspaces" class="issue-link" data-issue-key="JENKINS-4501"><del>JENKINS-4501</del></a>. As requested in <a href="https://issues.jenkins.io/browse/JENKINS-4501" title="WorkspaceCleanupThread deletes Matrix project workspaces" class="issue-link" data-issue-key="JENKINS-4501"><del>JENKINS-4501</del></a>, I am creating a new issue as this problem still exists in 1.602.</p>

<p>In short, Jenkins silently and erroneously deletes workspaces on slaves for matrix projects that are not old.</p>

<p>Over the course of the time I've worked with Jenkins this behavior has created literally days of work and waiting on very long running builds that rely on cached workspaces to be manageable. It's cost me more hours again today after restoring jenkins to a new server after a hardware failure. This setting was reset since it exists outside normal recommended backup files and I didn't think to add it when I "fixed" this last time.</p>

<p>Would it not be easier to have hudson.model.WorkspaceCleanupThread.disabled default to true? Having the default behavior be "destroy my data" seems bad, especially with how cheap disk is now. I'm sure when this option was implemented it made a lot of sense, but when I can get a 1TB for $50, it just seems wrong-headed. Let the fallow workspaces lie. I can clean them up if I need to.</p>

<p>If that's not an acceptable solution, could it not be moved to a config location in the Jenkins home? That way we can be relatively sure that the setting will be propagated in backups and not bite someone who thought they solved this problem and had forgotten about it?</p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Linux host, Linux, OSX, Windows, slaves. Jenkins version 1.602.</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>mwebber's <a href="https://issues.jenkins.io/browse/JENKINS-27329?focusedId=234136&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-234136">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>&gt;&gt; Jenkins silently and erroneously deletes workspaces on slaves for matrix projects that are not old
We just hit this problem (or what appears to be this problem) as well.</p>

<p>An extract from <tt>$JENKINS_HOME/Workspace clean-up.log</tt>:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>Deleting /Users/dlshudson/jenkins_slave/workspace/dials_distribute on dials-mac-mini
Deleting /scratch/jenkins_slave/workspace/dials_distribute on dials-ws133
Deleting /scratch/jenkins_slave/workspace/dials_distribute on dials-ws154
</pre>
</div></div>
<p><em>side-note</em>: it's a shame those log lines are not time-stamped
The 3 mentioned workspaces are from a matrix project, and all workspaces had been accessed recently.</p>

<p>Presumably there is a bug in the workspace cleanup code that means it does not handle matrix projects correctly.</p>

<p>Note that the job configuration specifies 4 slaves: 1 by label, and 3 by individual nodes. The workspaces that were deleted were those on the 3 slaves that were specified as individual notes, but the workspace on the slave that was specified by label was not deleted. Possibly a clue o the bug?</p>

<p>The workaround is to set <tt>hudson.model.WorkspaceCleanupThread.disabled=true</tt>.</p>
</pre>
</details>

</details>
<p>>> Jenkins silently and erroneously deletes workspaces on slaves for matrix projects that are not old<br/>
We just hit this problem (or what appears to be this problem) as well.</p>

<p>An extract from <tt>$JENKINS_HOME/Workspace clean-up.log</tt>:</p>


```
Deleting /Users/dlshudson/jenkins_slave/workspace/dials_distribute on dials-mac-mini
Deleting /scratch/jenkins_slave/workspace/dials_distribute on dials-ws133
Deleting /scratch/jenkins_slave/workspace/dials_distribute on dials-ws154
```
<p><em>side-note</em>: it's a shame those log lines are not time-stamped<br/>
The 3 mentioned workspaces are from a matrix project, and all workspaces had been accessed recently.</p>

<p>Presumably there is a bug in the workspace cleanup code that means it does not handle matrix projects correctly.</p>

<p>Note that the job configuration specifies 4 slaves: 1 by label, and 3 by individual nodes. The workspaces that were deleted were those on the 3 slaves that were specified as individual notes, but the workspace on the slave that was specified by label was not deleted. Possibly a clue o the bug?</p>

<p>The workaround is to set <tt>hudson.model.WorkspaceCleanupThread.disabled=true</tt>.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>mwebber's <a href="https://issues.jenkins.io/browse/JENKINS-27329?focusedId=234137&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-234137">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Daniel knows about this area, so assigning to him for comment (sorry, Daniel!)</p>
</pre>
</details>

</details>
<p>Daniel knows about this area, so assigning to him for comment (sorry, Daniel!)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/adamhong.jpg" title="adamhong's avatar" /> adamhong's <a href="https://issues.jenkins.io/browse/JENKINS-27329?focusedId=334066&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-334066">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>also seeing this happen recently, version 2.60.3</p>
</pre>
</details>

</details>
<p>also seeing this happen recently, version 2.60.3</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/rupunzlkim.svg" title="rupunzlkim's avatar" /> rupunzlkim's <a href="https://issues.jenkins.io/browse/JENKINS-27329?focusedId=334331&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-334331">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Sorry for the delay.  The version that we've noticed this on is 2.7.4</p>
</pre>
</details>

</details>
<p>Sorry for the delay.  The version that we've noticed this on is 2.7.4</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link's <a href="https://issues.jenkins.io/browse/JENKINS-27329?focusedId=338900&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-338900">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Code changed in jenkins
User: Reinhold Füreder
Path:
 core/src/main/java/hudson/model/WorkspaceCleanupThread.java
<a href="http://jenkins-ci.org/commit/jenkins/f258aff7a736a81306ecb7d3c56cacc9b3a09a68" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/f258aff7a736a81306ecb7d3c56cacc9b3a09a68</a>
Log:
  <a href="https://issues.jenkins.io/browse/JENKINS-27329" title="WorkspaceCleanupThread may delete workspaces of running jobs" class="issue-link" data-issue-key="JENKINS-27329"><del>JENKINS-27329</del></a> Less aggressive WorkspaceCleanupThread (#3444)</p>

<ul>
	<li><a href="https://issues.jenkins.io/browse/JENKINS-27329" title="WorkspaceCleanupThread may delete workspaces of running jobs" class="issue-link" data-issue-key="JENKINS-27329"><del>JENKINS-27329</del></a> Less aggressive WorkspaceCleanupThread</li>
</ul>


<p>I dare to claim that the default behaviour of WorkspaceCleanupThread is too aggressive =&gt; this little change is by no means perfect (or admittedly even far from perfect), but IMHO a saner or slightly more defensive default behaviour.</p>

<p>Mind that according to <a href="https://github.com/jenkinsci/jenkins/blob/9e64bcdcb4a2cf12d59dfa334e09ffb448d361e9/core/src/main/java/hudson/model/Job.java#L301" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/blob/9e64bcdcb4a2cf12d59dfa334e09ffb448d361e9/core/src/main/java/hudson/model/Job.java#L301</a> this "only" checks whether or not the last build of a job is in progress, while the JavaDoc says "Returns true if a build of this project is in progress." (cf. <a href="http://javadoc.jenkins-ci.org/hudson/model/Job.html#isBuilding--" class="external-link" target="_blank" rel="nofollow noopener">http://javadoc.jenkins-ci.org/hudson/model/Job.html#isBuilding--</a>)</p>

<ul>
	<li>Fix compilation</li>
</ul>


<ul>
	<li>Dummy commit to trigger pipeline</li>
</ul>


<p>Previous pipeline execution (<a href="https://ci.jenkins.io/blue/organizations/jenkins/Core%2Fjenkins/detail/PR-3444/2/tests" class="external-link" target="_blank" rel="nofollow noopener">https://ci.jenkins.io/blue/organizations/jenkins/Core%2Fjenkins/detail/PR-3444/2/tests</a>) failed with one failing test that at first glance appears to be unrelated with my change(s) and looks like a flaky test?</p>

<ul>
	<li>Add fine logging message</li>
</ul>




<p>      *<b>NOTE:</b>* This service been marked for deprecation: <a href="https://developer.github.com/changes/2018-04-25-github-services-deprecation/" class="external-link" target="_blank" rel="nofollow noopener">https://developer.github.com/changes/2018-04-25-github-services-deprecation/</a></p>

<p>      Functionality will be removed from GitHub.com on January 31st, 2019.</p>

</pre>
</details>

</details>
<p>Code changed in jenkins<br/>
User: Reinhold Füreder<br/>
Path:<br/>
 core/src/main/java/hudson/model/WorkspaceCleanupThread.java<br/>
<a href="http://jenkins-ci.org/commit/jenkins/f258aff7a736a81306ecb7d3c56cacc9b3a09a68" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/f258aff7a736a81306ecb7d3c56cacc9b3a09a68</a><br/>
Log:<br/>
  <a href="https://issues.jenkins.io/browse/JENKINS-27329" title="WorkspaceCleanupThread may delete workspaces of running jobs" class="issue-link" data-issue-key="JENKINS-27329"><del>JENKINS-27329</del></a> Less aggressive WorkspaceCleanupThread (#3444)</p>

<ul>
	<li><a href="https://issues.jenkins.io/browse/JENKINS-27329" title="WorkspaceCleanupThread may delete workspaces of running jobs" class="issue-link" data-issue-key="JENKINS-27329"><del>JENKINS-27329</del></a> Less aggressive WorkspaceCleanupThread</li>
</ul>


<p>I dare to claim that the default behaviour of WorkspaceCleanupThread is too aggressive => this little change is by no means perfect (or admittedly even far from perfect), but IMHO a saner or slightly more defensive default behaviour.</p>

<p>Mind that according to <a href="https://github.com/jenkinsci/jenkins/blob/9e64bcdcb4a2cf12d59dfa334e09ffb448d361e9/core/src/main/java/hudson/model/Job.java#L301" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/blob/9e64bcdcb4a2cf12d59dfa334e09ffb448d361e9/core/src/main/java/hudson/model/Job.java#L301</a> this "only" checks whether or not the last build of a job is in progress, while the JavaDoc says "Returns true if a build of this project is in progress." (cf. <a href="http://javadoc.jenkins-ci.org/hudson/model/Job.html#isBuilding--" class="external-link" target="_blank" rel="nofollow noopener">http://javadoc.jenkins-ci.org/hudson/model/Job.html#isBuilding--</a>)</p>

<ul>
	<li>Fix compilation</li>
</ul>


<ul>
	<li>Dummy commit to trigger pipeline</li>
</ul>


<p>Previous pipeline execution (<a href="https://ci.jenkins.io/blue/organizations/jenkins/Core%2Fjenkins/detail/PR-3444/2/tests" class="external-link" target="_blank" rel="nofollow noopener">https://ci.jenkins.io/blue/organizations/jenkins/Core%2Fjenkins/detail/PR-3444/2/tests</a>) failed with one failing test that at first glance appears to be unrelated with my change(s) and looks like a flaky test?</p>

<ul>
	<li>Add fine logging message</li>
</ul>




<p>      *<b>NOTE:</b>* This service been marked for deprecation: <a href="https://developer.github.com/changes/2018-04-25-github-services-deprecation/" class="external-link" target="_blank" rel="nofollow noopener">https://developer.github.com/changes/2018-04-25-github-services-deprecation/</a></p>

<p>      Functionality will be removed from GitHub.com on January 31st, 2019.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/oleg_nenashev.png" title="oleg_nenashev's avatar" /> oleg_nenashev's <a href="https://issues.jenkins.io/browse/JENKINS-27329?focusedId=339896&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-339896">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Fix has been applied in 2.125. IMHO the fix is not complete for parallel AbstractProject builds, but it is better than nothing. Will create a follow-up ticket</p>
</pre>
</details>

</details>
<p>Fix has been applied in 2.125. IMHO the fix is not complete for parallel AbstractProject builds, but it is better than nothing. Will create a follow-up ticket</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/oleg_nenashev.png" title="oleg_nenashev's avatar" /> oleg_nenashev's <a href="https://issues.jenkins.io/browse/JENKINS-27329?focusedId=339897&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-339897">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=danielbeck" class="user-hover" rel="danielbeck">danielbeck</a> this thing is marked as RFE in the changelog, but I think this is a bug. Would you agree if I recategorize it?</p>
</pre>
</details>

</details>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=danielbeck" class="user-hover" rel="danielbeck">danielbeck</a> this thing is marked as RFE in the changelog, but I think this is a bug. Would you agree if I recategorize it?</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Originally duplicated by: <a href="https://github.com/timja-org/jenkins-gh-issues-poc-11-15/issues?q=in%3Atitle%20JENKINS-30916">JENKINS-30916</a>]</i>
